components/normalizer/src/provider.rs

```diff
@@ -42,101 +42,51 @@ const _: () = {
     CanonicalCompositionsV1Marker::INFO,
-    CanonicalDecompositionDataV1Marker::INFO,
+    CanonicalDecompositionDataV2Marker::INFO,
     CanonicalDecompositionTablesV1Marker::INFO,
-    CompatibilityDecompositionSupplementV1Marker::INFO,
+    CompatibilityDecompositionDataV2Marker::INFO,
     CompatibilityDecompositionTablesV1Marker::INFO,
     NonRecursiveDecompositionSupplementV1Marker::INFO,
-    Uts46DecompositionSupplementV1Marker::INFO,
+    Uts46DecompositionDataV2Marker::INFO,
 ];
 
-/// Main data for NFD
-///
-/// <div class="stab unstable">
-/// 🚧 This code is considered unstable; it may change at any time, in breaking or non-breaking ways,
-/// including in SemVer minor releases. While the serde representation of data structs is guaranteed
-/// to be stable, their Rust representation might not be. Use with caution.
-/// </div>
-#[icu_provider::data_struct(marker(
-    CanonicalDecompositionDataV1Marker,
-    "normalizer/nfd@1",
-    singleton
-))]
-#[derive(Debug, PartialEq, Clone)]
-#[cfg_attr(feature = "datagen", derive(serde::Serialize, databake::Bake))]
-#[cfg_attr(feature = "datagen", databake(path = icu_normalizer::provider))]
-#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
-pub struct DecompositionDataV1<'data> {
-    /// Trie for NFD decomposition.
-    #[cfg_attr(feature = "serde", serde(borrow))]
-    pub trie: CodePointTrie<'data, u32>,
-}
-
-/// Data that either NFKD or the decomposed form of UTS 46 needs
-/// _in addition to_ the NFD data.
+/// Decomposition data
 ///
 /// <div class="stab unstable">
 /// 🚧 This code is considered unstable; it may change at any time, in breaking or non-breaking ways,
 /// including in SemVer minor releases. While the serde representation of data structs is guaranteed
 /// to be stable, their Rust representation might not be. Use with caution.
 /// </div>
 #[icu_provider::data_struct(
-    marker(
-        CompatibilityDecompositionSupplementV1Marker,
-        "normalizer/nfkd@1",
-        singleton
-    ),
-    marker(Uts46DecompositionSupplementV1Marker, "normalizer/uts46d@1", singleton)
+    marker(CanonicalDecompositionDataV2Marker, "normalizer/nfd@2", singleton),
+    marker(CompatibilityDecompositionDataV2Marker, "normalizer/nfkd@2", singleton),
+    marker(Uts46DecompositionDataV2Marker, "normalizer/uts46d@2", singleton)
 )]
 #[derive(Debug, PartialEq, Clone)]
 #[cfg_attr(feature = "datagen", derive(serde::Serialize, databake::Bake))]
 #[cfg_attr(feature = "datagen", databake(path = icu_normalizer::provider))]
 #[cfg_attr(feature = "serde", derive(serde::Deserialize))]
-pub struct DecompositionSupplementV1<'data> {
-    /// Trie for the decompositions that differ from NFD.
-    /// Getting a zero from this trie means that you need
-    /// to make another lookup from `DecompositionDataV1::trie`.
+pub struct DecompositionDataV2<'data> {
+    /// Trie for decomposition.
     #[cfg_attr(feature = "serde", serde(borrow))]
     pub trie: CodePointTrie<'data, u32>,
```

Comment:
issue: I feel like the packed code logic is all scattered. Can we use a structured `NormalizationTrieValue(pub u32)` type that has convenience methods for getting all the fields?

Reply:
I agree that what you suggest would be better for encapsulation. However, given that prior to this PR there was no such encapsulation and I'm already way over my time budget for this, I would very much prefer landing this ASAP (before 2.0 and before this bitrots) without such a refactoring and leaving the refactoring as a follow-up.